db_setup/app/__init__.py

```diff
@@ -16,63 +17,137 @@
+    unique_resources = []
+    resource_dict = {}
+    for resource in data:
+        if not resource_dict.get(resource['url']):
+            resource_dict[resource['url']] = True
+            unique_resources.append(resource)
+    return unique_resources
+
+def get_category(resource, category_dict):
+    category = resource.get('category')
+
+    if category not in category_dict:
+        category_dict[category] = Category(name=category)
+
+    return category_dict[category]
+
+def get_languages(resource, language_dict):
+    langs = []
+
+    # Loop through languages and create a new Language
+    # object for any that don't exist in the DB
+    for language in resource.get('languages') or []:
+        if language not in language_dict:
+            language_dict[language] = Language(name=language)
+
+        # Add each Language object associated with this resource
+        # to the list we'll return
+        langs.append(language_dict[language])
+
+    return langs
+
+def create_resource(resource, langs, category):
+    try:
+        new_resource = Resource(
+            name=resource['name'],
+            url=resource['url'],
+            category=category,
+            languages=langs,
+            paid=resource.get('paid'),
+            notes=resource.get('notes', ''),
+            upvotes=resource.get('upvotes', 0),
+            downvotes=resource.get('downvotes', 0),
+            times_clicked=resource.get('times_clicked', 0))
+
+        db.session.add(new_resource)
+        db.session.commit()
+    except exc.SQLAlchemyError as e:
+        db.session.rollback()
+        print('Flask SQLAlchemy Exception:', e)
+        template = "An SQLAlchemy exception of type {0} occurred. Arguments:\n{1!r}"
+        print(resource)
+    except Exception as e:
+        db.session.rollback()
+        print('exception', e)
+        template = "An exception of type {0} occurred. Arguments:\n{1!r}"
+        print(resource)
+
+def update_resource(resource, existing_resource, langs, category):
+    # Return without updating if the resource is already up to date
+    if match_resource(resource, existing_resource, langs):
+        return
+
+    try:
+        existing_resource.name = resource['name']
+        existing_resource.url = resource['url']
+        existing_resource.category = category
+        existing_resource.paid = resource.get('paid')
+        existing_resource.notes = resource.get('notes', '')
+        existing_resource.languages = langs
+
+        db.session.commit()
+    except exc.SQLAlchemyError as e:
+        db.session.rollback()
+        print('Flask SQLAlchemy Exception:', e)
+        template = "An SQLAlchemy exception of type {0} occurred. Arguments:\n{1!r}"
+        print(resource)
+    except Exception as e:
+        db.session.rollback()
+        print('exception', e)
+        template = "An exception of type {0} occurred. Arguments:\n{1!r}"
+        print(resource)
+
+def match_resource(resource_dict, resource_obj, langs):
```

Comment:
this comparison can be moved into the `Resource` class, using the `__eq__` magic method, or keep it as is. 


```
def match_resource(resource_dict, resource_obj):
    comparable_attributes = ['name', 'paid', 'notes', 'category', 'languages']
    
    return all( resource_dict.get(attr_type) == getattr(resource_obj, attr_type) for attr_type in comparable_attributes)
```


this can be done when you set the `__eq__` and `__hash__` methods for category also (like you did for languages).

Reply:
I like this, I'm definitely going to move this into the `Resource` class and define `__eq__` and `__hash__` for Resource and Category